src/algebra/star/star_alg_hom.lean

```diff
@@ -0,0 +1,445 @@
+
+/-- The `non_unital_star_alg_hom.prod` of two morphisms is a morphism. -/
+@[simps] def prod (f : A →⋆ₙₐ[R] B) (g : A →⋆ₙₐ[R] C) : (A →⋆ₙₐ[R] B × C) :=
+{ to_fun    := pi.prod f g,
+  map_zero' := by simp only [pi.prod, prod.zero_eq_mk, map_zero],
+  map_add'  := λ x y, by simp only [pi.prod, prod.mk_add_mk, map_add],
+  map_mul'  := λ x y, by simp only [pi.prod, prod.mk_mul_mk, map_mul],
+  map_smul' := λ c x, by simp only [pi.prod, prod.smul_mk, map_smul, ring_hom.id_apply],
+  map_star' := λ x, by simp only [pi.prod, prod.star_def, map_star] }
+
+lemma coe_prod (f : A →ₙₐ[R] B) (g : A →ₙₐ[R] C) : ⇑(f.prod g) = pi.prod f g := rfl
+
+@[simp] theorem fst_prod (f : A →⋆ₙₐ[R] B) (g : A →⋆ₙₐ[R] C) :
+  (fst R B C).comp (prod f g) = f := by ext; refl
+
+@[simp] theorem snd_prod (f : A →⋆ₙₐ[R] B) (g : A →⋆ₙₐ[R] C) :
+  (snd R B C).comp (prod f g) = g := by ext; refl
+
+@[simp] theorem prod_fst_snd : prod (fst R A B) (snd R A B) = 1 :=
+fun_like.coe_injective pi.prod_fst_snd
+
+/-- Taking the product of two maps with the same domain is equivalent to taking the product of
+their codomains. -/
+@[simps] def prod_equiv : ((A →⋆ₙₐ[R] B) × (A →⋆ₙₐ[R] C)) ≃ (A →⋆ₙₐ[R] B × C) :=
+{ to_fun := λ f, f.1.prod f.2,
+  inv_fun := λ f, ((fst _ _ _).comp f, (snd _ _ _).comp f),
+  left_inv := λ f, by ext; refl,
+  right_inv := λ f, by ext; refl }
+
+end prod
+
+section inl_inr
+
+variables (R A B C : Type*) [monoid R]
+  [non_unital_non_assoc_semiring A] [distrib_mul_action R A] [star_add_monoid A]
+  [non_unital_non_assoc_semiring B] [distrib_mul_action R B] [star_add_monoid B]
+  [non_unital_non_assoc_semiring C] [distrib_mul_action R C] [star_add_monoid C]
+
+/-- The left injection into a product is a non-unital algebra homomorphism. -/
+def inl : A →⋆ₙₐ[R] A × B := prod 1 0
+
+/-- The right injection into a product is a non-unital algebra homomorphism. -/
+def inr : B →⋆ₙₐ[R] A × B := prod 0 1
+
+variables {R A B}
+
+@[simp] theorem coe_inl : (inl R A B : A → A × B) = λ x, (x, 0) := rfl
+theorem inl_apply (x : A) : inl R A B x = (x, 0) := rfl
+
+@[simp] theorem coe_inr : (inr R A B : B → A × B) = prod.mk 0 := rfl
+theorem inr_apply (x : B) : inr R A B x = (0, x) := rfl
+
+end inl_inr
+
+end non_unital_star_alg_hom
+
+namespace star_alg_hom
+
+variables (R A B C : Type*) [comm_semiring R]
+  [semiring A] [algebra R A] [has_star A]
+  [semiring B] [algebra R B] [has_star B]
+  [semiring C] [algebra R C] [has_star C]
+
+/-- The first projection of a product is a ⋆-algebra homomoprhism. -/
+@[simps]
+def fst : A × B →⋆ₐ[R] A :=
+{ to_fun := prod.fst, map_zero' := rfl, map_add' := λ x y, rfl, map_one' := rfl,
+  commutes' := λ x, rfl, map_mul' := λ x y, rfl, map_star' := λ x, rfl }
+
+/-- The second projection of a product is a ⋆-algebra homomorphism. -/
+@[simps]
+def snd : A × B →⋆ₐ[R] B :=
+{ to_fun := prod.snd, map_zero' := rfl, map_add' := λ x y, rfl, map_one' := rfl,
+  commutes' := λ x, rfl, map_mul' := λ x y, rfl, map_star' := λ x, rfl }
```

Comment:
Do we have these already for `alg_hom`? Can you reuse the construction for all but the new `map_star'` field?

Same below for `prod`.

Reply:
Looks like we have `fst` and `snd`, but not `prod`. Do you want me to add `alg_hom.prod` first?

Reply:
#16116